README.md
# Responsive NavBar!

this package delivers you a navigationBar with animation and much customization!

## Full Mobile Support
![](https://github.com/bknig99/responsive_navbar/blob/master/readme_images/mobile.png)

## Full Desktop Support
![](https://github.com/bknig99/responsive_navbar/blob/master/readme_images/desktop.png)

## Usage
```dart
@override
Widget build(BuildContext context) {
  return Container(
      child: ResponsiveBar(
          barCircularity: 0,
          iconPadding: EdgeInsets.only(bottom: 5, top: 5),
          barThickness: 5,
          usingRightBar: true,
          isForDesktop: true,
          currentIndex: currentIndex,
          onTap: _handleIndexChanged(),
          iconSize: 35,
          scaling: 60,
          iconScaleAnimationFactor: 0.3,
          iconScaleCurve: Curves.decelerate,
          barSizeCurve: Curves.decelerate,
          barAccentColor: FlutterFlowTheme.primaryColor,
          backgroundColor: HexColor("#060606"),
          activeItemColor: HexColor("#F1F1F1"),
          passiveItemColor: HexColor("#444444"),
          items: [
            ResponsiveBarItem(
                icon: Icon(Icons.store_outlined),
                selectedIcon: Icon(Icons.store_rounded)),
            ResponsiveBarItem(
                icon: Icon(Icons.badge_outlined),
                selectedIcon: Icon(Icons.badge_rounded)),
            ResponsiveBarItem(
                icon: Icon(Icons.vpn_key_outlined),
                selectedIcon: Icon(Icons.vpn_key_rounded)),
            ResponsiveBarItem(
                icon: Icon(Icons.source_outlined),
                selectedIcon: Icon(Icons.source_rounded)),
            ResponsiveBarItem(
                icon: Icon(Icons.linear_scale_outlined),
                selectedIcon: Icon(Icons.linear_scale_rounded))
          ]
      )
  );
}
```


CHANGELOG.md
## 0.0.8 fix icon animation

## 0.0.7 added tooltip functionality to items

## 0.0.6 cleanup code

## 0.0.5 fix usage problem with ResponsiveBarItem

## 0.0.4 add itemsStaked property 

## 0.0.3 fix null safety warning

## 0.0.2 fix usage problem with ResponsiveBarItem

## 0.0.1 Added general navigationbar

flutter_responsive_example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility in the flutter_test package. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_responsive_example/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/badge_text.dart
library responsive_navbar;

import 'package:flutter/material.dart';

///
/// @author bknig99
///
//##############################################################################
class BadgeText extends StatelessWidget {
  final int count;

  final double right;

  final bool show;

  const BadgeText({
    Key? key,
    required this.count,
    required this.right,
    required this.show,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Positioned(
      right: right,
      top: 0,
      child: show
          ? Container(
              padding: const EdgeInsets.all(2),
              decoration: BoxDecoration(
                color: Colors.red,
                borderRadius: BorderRadius.circular(6),
              ),
              constraints: const BoxConstraints(
                minWidth: 12,
                minHeight: 12,
              ),
              child: Text(
                count > 10 ? '10+' : '$count',
                style: const TextStyle(
                  color: Colors.white,
                  fontSize: 6,
                ),
                textAlign: TextAlign.center,
              ),
            )
          : Container(),
    );
  }
}


lib/responsive_navbar.dart
library responsive_navbar;

import 'package:flutter/material.dart';
import 'package:responsive_navbar/align_type.dart';
import 'package:responsive_navbar/badge_text.dart';

///
/// @author bknig99
///
//##############################################################################
class ResponsiveBar extends StatefulWidget {
  final bool usingBottomBar;
  final bool isForDesktop;
  final bool usingTopBar;
  final bool usingLeftBar;
  final bool usingRightBar;

  final bool usingItemDescription;
  final Duration tooltipStayDuration;
  final Duration tooltipWaitDuration;

  ///
  ///sets the roundness of the Bars
  ///################################
  ///default is 25
  ///set this param to 0 if you dont
  ///want to have roundness on the Bars
  final double barCircularity;

  ///
  /// sets the thickness of all Bars
  final double barThickness;

  ///
  /// adds Padding to the Icon
  /// ------------------------
  /// very useful in combination
  /// with the borders
  final EdgeInsets iconPadding;

  ///
  ///Color of the Sidebar itself
  final Color backgroundColor;

  ///
  ///Color of the selected Icons && Bars
  final Color activeItemColor;

  ///
  ///Color of the unselected Icons
  final Color passiveItemColor;

  ///
  ///Scaling for Sidebar
  ///--------------------------
  ///Desktop: width for Sidebar
  ///Mobile: height for Sidebar
  ///-------------------------
  final double scaling;

  ///
  ///Animationmultiplier for Icon Animation
  ///-------------------------
  /// Value > 0; && Value < 0.5
  ///-------------------------
  final double iconScaleAnimationFactor;

  ///
  ///List of the SidebarItems
  ///-> Icon, title etc are hold by them
  final List<ResponsiveBarItem> items;

  ///
  /// sets the navigationbutton layout
  /// false equals "expanding items over whole navbar"
  /// ##############################
  /// default = false
  final bool itemsStaked;

  ///
  ///index of the current Page
  final int currentIndex;

  ///
  ///Size for the Icon
  final double iconSize;

  ///
  ///Animation for Icon
  final Curve iconScaleCurve;

  ///
  /// if true, animation for Icons
  /// can be set with [iconScaleCurve] param
  final bool useIconScaleCurve;

  ///
  ///Animation for all activated Bars
  final Curve barSizeCurve;

  ///
  /// if true, animation for Bars
  /// can be set with [barSizeCurve] param
  final bool useBarAnimation;

  ///
  ///Amount of time all the animations will take
  final Duration transition;

  ///
  ///Callback function for pressing one Item
  final Function(int) onTap;

  ///
  ///Accent color for item
  ///-------------------------
  ///This color finishes the animation
  ///by doing color transition to bars
  ///-------------------------
  final Color barAccentColor;

  ///
  /// default is true,
  /// if you wish no Bars, set this to false
  /// ######################################
  /// otherwise, your prefered Color
  /// can be set with [barAccentColor] param
  final bool useBarAccentColor;

  const ResponsiveBar(
      {Key? key,
      required this.isForDesktop,
      this.scaling = 60,
      this.activeItemColor = const Color(0xFFFFFFFF),
      this.passiveItemColor = const Color(0xAAAAAAAA),
      required this.items,
      this.iconScaleAnimationFactor = 0.2,
      required this.onTap,
      this.currentIndex = 0,
      this.iconSize = 24.0,
      this.backgroundColor = Colors.white,
      this.barSizeCurve = Curves.linear,
      required this.iconScaleCurve,
      this.barAccentColor = const Color(0xFFFFFFFF),
      this.transition = const Duration(milliseconds: 750),
      this.usingTopBar = false,
      this.usingBottomBar = false,
      this.usingLeftBar = false,
      this.usingRightBar = false,
      this.iconPadding = EdgeInsets.zero,
      this.barThickness = 4,
      this.barCircularity = 25,
      this.useBarAnimation = false,
      this.useBarAccentColor = true,
      this.useIconScaleCurve = true,
      this.itemsStaked = false,
      this.usingItemDescription = false,
      this.tooltipStayDuration = const Duration(milliseconds: 500),
      this.tooltipWaitDuration = const Duration(milliseconds: 500)})
      : assert(iconScaleAnimationFactor <= 0.5,
            'Scale factor must be smaller than 0.5'),
        assert(
            iconScaleAnimationFactor > 0, 'Scale factor must be bigger than 0'),
        assert(0 <= currentIndex && currentIndex < items.length),
        super(key: key);

  @override
  ResponsiveBarState createState() => ResponsiveBarState();
}

//##############################################################################
//_SideBarState
//##############################################################################
class ResponsiveBarState extends State<ResponsiveBar>
    with TickerProviderStateMixin {
  final ValueNotifier<List<double>> _sizes = ValueNotifier([]);
  late AnimationController _scaleIconController;
  late List<Color> _barColorList;
  late List<BarsOfItem> _barsOfItemList;

  @override
  void initState() {
    super.initState();

    _barsOfItemList = List<BarsOfItem>.generate(widget.items.length, (index) {
      return BarsOfItem(Bar(0, 0), Bar(0, 0), Bar(0, 0), Bar(0, 0));
    });
    _sizes.value = List<double>.generate(widget.items.length, (index) {
      return 0;
    });
    _barColorList = List<Color>.generate(widget.items.length, (index) {
      return widget.activeItemColor;
    });

    //initial animation
    _animateBarWidth(widget.currentIndex, widget.scaling * 0.65);
    _animateBarHeight(widget.currentIndex, 3.5);
    _animateBarColor(widget.currentIndex, widget.barAccentColor);
    _animateIconScale(widget.currentIndex);
  }

  @override
  void dispose() {
    _scaleIconController.dispose();
    super.dispose();
  }

  //############################################################################

  @override
  Widget build(BuildContext context) {
    final scaling = widget.scaling;

    return Material(
      color: widget.backgroundColor,
      child: widget.isForDesktop
          ? desktopContainer(scaling)
          : mobileContainer(scaling),
    );
  }

  Widget desktopContainer(scaling) {
    return SizedBox(
      width: scaling,
      height: MediaQuery.of(context).size.height,
      child: Column(
        crossAxisAlignment: widget.itemsStaked
            ? CrossAxisAlignment.start
            : CrossAxisAlignment.center,
        children: <Widget>[
          for (var i = 0; i < widget.items.length; i++)
            widget.itemsStaked
                ? GestureDetector(
                    behavior: HitTestBehavior.opaque,
                    onTap: () {
                      widget.onTap(i);
                    },
                    child: Column(
                      mainAxisAlignment: widget.itemsStaked
                          ? MainAxisAlignment.start
                          : MainAxisAlignment.center,
                      crossAxisAlignment: CrossAxisAlignment.center,
                      children: [
                        _buildNavigationItem(i),
                      ],
                    ))
                : Expanded(
                    child: GestureDetector(
                        behavior: HitTestBehavior.opaque,
                        onTap: () {
                          widget.onTap(i);
                        },
                        child: Column(
                          mainAxisAlignment: widget.itemsStaked
                              ? MainAxisAlignment.start
                              : MainAxisAlignment.center,
                          crossAxisAlignment: CrossAxisAlignment.center,
                          children: [
                            _buildNavigationItem(i),
                          ],
                        )),
                  ),
        ],
      ),
    );
  }

  Widget mobileContainer(scaling) {
    return SizedBox(
      width: MediaQuery.of(context).size.width,
      height: scaling,
      child: Row(
        crossAxisAlignment: widget.itemsStaked
            ? CrossAxisAlignment.start
            : CrossAxisAlignment.center,
        children: <Widget>[
          for (var i = 0; i < widget.items.length; i++)
            Expanded(
              child: GestureDetector(
                behavior: HitTestBehavior.opaque,
                onTap: () {
                  widget.onTap(i);
                },
                child: Row(
                  mainAxisAlignment: widget.itemsStaked
                      ? MainAxisAlignment.start
                      : MainAxisAlignment.center,
                  crossAxisAlignment: CrossAxisAlignment.center,
                  children: [
                    _buildNavigationItem(i),
                  ],
                ),
              ),
            ),
        ],
      ),
    );
  }

  Widget _buildNavigationItem(int index) {
    return Column(
        mainAxisAlignment: MainAxisAlignment.center,
        crossAxisAlignment: CrossAxisAlignment.center,
        children: [
          //topBorder
          widget.usingTopBar ? _resolveTopBorder(index) : Container(),

          //leftBorder - Icon - rightBorder
          widget.usingItemDescription
              ? Tooltip(
                  showDuration: widget.tooltipStayDuration,
                  waitDuration: widget.tooltipWaitDuration,
                  message: widget.items[index].description,
                  child: _buildIcon(index))
              : _buildIcon(index),

          //bottomBorder
          widget.usingBottomBar ? _resolveBottomBorder(index) : Container()
        ]);
  }

  //############################################################################
  //StaticBorder
  //############################################################################

  Widget _resolveStaticBorder(int index, AlignType alignType) {
    return Container(
      height: _barsOfItemList[index].getHeight(alignType),
      width: _barsOfItemList[index].getWidth(alignType),
      decoration: BoxDecoration(
          color: _barColorList[index],
          borderRadius:
              BorderRadius.all(Radius.circular(widget.barCircularity))),
    );
  }

  //############################################################################
  //TopBorder
  //############################################################################
  Widget _resolveTopBorder(int index) {
    return widget.useBarAnimation
        ? AnimatedContainer(
            height: _barsOfItemList[index].getHeight(AlignType.TOP),
            width: _barsOfItemList[index].getWidth(AlignType.TOP),
            duration: widget.transition,
            curve: widget.barSizeCurve,
            decoration: BoxDecoration(
                color: _barColorList[index],
                borderRadius:
                    BorderRadius.all(Radius.circular(widget.barCircularity))),
          )
        : _resolveStaticBorder(index, AlignType.TOP);
  }

  //############################################################################
  //Icon
  //############################################################################
  Widget _buildIcon(int index) {
    return Container(
      padding: EdgeInsets.only(
          top: widget.iconPadding.top, bottom: widget.iconPadding.bottom),
      child: Row(
        mainAxisAlignment: MainAxisAlignment.spaceBetween,
        children: [
          widget.usingLeftBar ? _resolveLeftBorder(index) : Container(),
          SizedBox(
            height: widget.iconSize,
            width: widget.iconSize,
            child: CustomPaint(
              child: _CustomNavigationBarTile(
                iconSize: widget.iconSize,
                scale: _sizes.value[index],
                selected: index == widget.currentIndex,
                item: widget.items[index],
                selectedColor: widget.activeItemColor,
                unSelectedColor: widget.passiveItemColor,
              ),
            ),
          ),
          widget.usingRightBar ? _resolveRightBorder(index) : Container()
        ],
      ),
    );
  }

  //############################################################################
  //BottomBorder
  //############################################################################
  Widget _resolveBottomBorder(int index) {
    return widget.useBarAnimation
        ? AnimatedContainer(
            height: _barsOfItemList[index].getHeight(AlignType.BOTTOM),
            width: _barsOfItemList[index].getWidth(AlignType.BOTTOM),
            duration: widget.transition,
            curve: widget.barSizeCurve,
            decoration: BoxDecoration(
                color: _barColorList[index],
                borderRadius:
                    BorderRadius.all(Radius.circular(widget.barCircularity))),
          )
        : _resolveStaticBorder(index, AlignType.BOTTOM);
  }

  //############################################################################
  //LeftBorder
  //############################################################################

  Widget _resolveLeftBorder(int index) {
    return widget.useBarAnimation
        ? AnimatedContainer(
            height: _barsOfItemList[index].getHeight(AlignType.LEFT),
            width: _barsOfItemList[index].getWidth(AlignType.LEFT),
            duration: widget.transition,
            curve: widget.barSizeCurve,
            decoration: BoxDecoration(
                color: _barColorList[index],
                borderRadius:
                    BorderRadius.all(Radius.circular(widget.barCircularity))),
          )
        : Container(
            height: _barsOfItemList[index].getHeight(AlignType.LEFT),
            width: _barsOfItemList[index].getWidth(AlignType.LEFT),
            decoration: BoxDecoration(
                color: _barColorList[index],
                borderRadius:
                    BorderRadius.all(Radius.circular(widget.barCircularity))),
          );
  }

  //############################################################################
  //RightBorder
  //############################################################################

  Widget _resolveRightBorder(int index) {
    return widget.useBarAnimation
        ? AnimatedContainer(
            height: _barsOfItemList[index].getHeight(AlignType.RIGHT),
            width: _barsOfItemList[index].getWidth(AlignType.RIGHT),
            duration: widget.transition,
            curve: widget.barSizeCurve,
            decoration: BoxDecoration(
                color: _barColorList[index],
                borderRadius:
                    BorderRadius.all(Radius.circular(widget.barCircularity))),
          )
        : Container(
            height: _barsOfItemList[index].getHeight(AlignType.RIGHT),
            width: _barsOfItemList[index].getWidth(AlignType.RIGHT),
            decoration: BoxDecoration(
                color: _barColorList[index],
                borderRadius:
                    BorderRadius.all(Radius.circular(widget.barCircularity))),
          );
  }

  //############################################################################
  // start and define all animations
  //############################################################################
  @override
  void didUpdateWidget(ResponsiveBar oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (widget.currentIndex != oldWidget.currentIndex) {
      _animateBarWidth(widget.currentIndex,
          widget.scaling * (widget.isForDesktop ? 0.75 : 0.6));
      _animateBarWidth(oldWidget.currentIndex, 0);
      _animateBarHeight(widget.currentIndex, widget.barThickness);
      _animateBarHeight(oldWidget.currentIndex, 0);
      if (widget.useBarAccentColor) {
        _animateBarColor(widget.currentIndex, widget.barAccentColor);
        _animateBarColor(oldWidget.currentIndex, widget.activeItemColor);
      }
      _animateIconScale(widget.currentIndex);
    }
  }

  //######################################

  _animateBarWidth(int index, double length) {
    _barsOfItemList[index].setAllWidths(length);
  }

  //######################################

  _animateBarHeight(int index, double length) {
    _barsOfItemList[index].setAllHeights(length);
  }

  //######################################

  void _animateBarColor(int index, Color color) {
    _barColorList[index] = color;
  }

  //######################################

  void _animateIconScale(int index) {
    if (widget.useIconScaleCurve) {
      _scaleIconController = AnimationController(
          vsync: this,
          duration:
              Duration(milliseconds: widget.transition.inMilliseconds ~/ 2))
        ..addStatusListener((status) {
          if (status == AnimationStatus.completed) {
            _scaleIconController.reverse();
          }
        });

      CurvedAnimation scaleAnimation = CurvedAnimation(
        parent: _scaleIconController,
        curve: widget.iconScaleCurve,
        reverseCurve: widget.iconScaleCurve.flipped,
      );

      Tween<double>(begin: 0.0, end: 1.0)
          .animate(scaleAnimation)
          .addListener(() {
        setState(() {
          _sizes.value[index] =
              scaleAnimation.value * widget.iconScaleAnimationFactor;
        });
      });
      if (widget.currentIndex == widget.currentIndex) {
          _scaleIconController.forward();
      }
    }
  }
}

//##############################################################################
//_CustomNavigationBarTile
//##############################################################################
class _CustomNavigationBarTile extends StatelessWidget {
  final bool selected;
  final ResponsiveBarItem item;
  final Color selectedColor;
  final Color unSelectedColor;
  final double scale;
  final double iconSize;

  //############################################################################

  const _CustomNavigationBarTile(
      {Key? key,
      required this.selected,
      required this.item,
      required this.selectedColor,
      required this.unSelectedColor,
      required this.scale,
      required this.iconSize})
      : super(key: key);

  //############################################################################

  @override
  Widget build(BuildContext context) {
    return Transform.scale(
      scale: 1.0 + scale,
      child: Stack(
        children: [
          IconTheme(
            data: IconThemeData(
              color: selected ? selectedColor : unSelectedColor,
              size: iconSize,
            ),
            child: selected ? item.selectedIcon : item.icon,
          ),
          BadgeText(
            show: item.showBadge,
            count: item.badgeCount,
            right: 0.0,
          )
        ],
      ),
    );
  }
}

//##############################################################################
//BarsOfItem (== 4 Bars for 1 Item in ResponsiveBar)
//##############################################################################
class BarsOfItem {
  Bar top;
  Bar bottom;
  Bar left;
  Bar right;

  BarsOfItem(this.top, this.bottom, this.left, this.right);

  // ignore: missing_return
  double getWidth(AlignType alignType) {
    switch (alignType) {
      case AlignType.TOP:
        return top.width;
      case AlignType.BOTTOM:
        return bottom.width;
      case AlignType.LEFT:
        return left.height;
      case AlignType.RIGHT:
        return right.height;
    }
  }

  double getHeight(AlignType alignType) {
    switch (alignType) {
      case AlignType.TOP:
        return top.height;
      case AlignType.BOTTOM:
        return bottom.height;
      case AlignType.LEFT:
        return left.width;
      case AlignType.RIGHT:
        return right.width;
    }
  }

  setAllWidths(double value) {
    top = Bar(value, top.height);
    bottom = Bar(value, bottom.height);
    left = Bar(value, left.width);
    right = Bar(value, right.width);
  }

  setAllHeights(double value) {
    top = Bar(top.width, value);
    bottom = Bar(bottom.width, value);
    left = Bar(left.width, value);
    right = Bar(right.width, value);
  }
}

//##############################################################################
//Bar
//##############################################################################
class Bar {
  double width;
  double height;

  Bar(this.width, this.height);
}

//##############################################################################
//ResponsiveBarItem
//##############################################################################
class ResponsiveBarItem {
  ///
  /// The icon of the item
  /// Typically the icon is of type [Icon]
  ///
  final Widget icon;

  final String description;

  /// An alternative icon displayed when this bottom navigation item is
  /// selected.
  /// ------------------------------------
  /// If this icon is not provided, the bottom navigation bar will display
  /// [icon] in either state.
  final Widget selectedIcon;

  /// Notification badge count
  final int badgeCount;

  /// hide or show badge
  final bool showBadge;

  /// Create a Custom Navigationbar Item
  ///
  /// the [selectedIcon] must not be null
  /// the [icon] must not be null
  ResponsiveBarItem(
      {required this.icon,
      required this.selectedIcon,
      this.badgeCount = 0,
      this.showBadge = false,
      this.description = ""});
}


flutter_responsive_example/windows/runner/main.cpp
#include <flutter/dart_project.h>
#include <flutter/flutter_view_controller.h>
#include <windows.h>

#include "flutter_window.h"
#include "utils.h"

int APIENTRY wWinMain(_In_ HINSTANCE instance, _In_opt_ HINSTANCE prev,
                      _In_ wchar_t *command_line, _In_ int show_command) {
  // Attach to console when present (e.g., 'flutter run') or create a
  // new console when running with a debugger.
  if (!::AttachConsole(ATTACH_PARENT_PROCESS) && ::IsDebuggerPresent()) {
    CreateAndAttachConsole();
  }

  // Initialize COM, so that it is available for use in the library and/or
  // plugins.
  ::CoInitializeEx(nullptr, COINIT_APARTMENTTHREADED);

  flutter::DartProject project(L"data");

  std::vector<std::string> command_line_arguments =
      GetCommandLineArguments();

  project.set_dart_entrypoint_arguments(std::move(command_line_arguments));

  FlutterWindow window(project);
  Win32Window::Point origin(10, 10);
  Win32Window::Size size(1280, 720);
  if (!window.Create(L"flutter_responsive_example", origin, size)) {
    return EXIT_FAILURE;
  }
  window.SetQuitOnClose(true);

  ::MSG msg;
  while (::GetMessage(&msg, nullptr, 0, 0)) {
    ::TranslateMessage(&msg);
    ::DispatchMessage(&msg);
  }

  ::CoUninitialize();
  return EXIT_SUCCESS;
}


example/lib/main.dart
// ignore_for_file: must_be_immutable

import 'package:flutter/material.dart';
import 'package:responsive_navbar/responsive_navbar.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: DesktopMain(title: 'Flutter Demo Home Page'),
    );
  }
}

class DesktopMain extends StatefulWidget {
  DesktopMain({super.key, required this.title, this.index = 0});

  final String title;

  late int index;

  @override
  State<DesktopMain> createState() => _DesktopMainState();
}

class _DesktopMainState extends State<DesktopMain> {
  final ValueNotifier<int> _currentIndex = ValueNotifier(0);
  late PageController pageController;

  @override
  void initState() {
    super.initState();
    pageController = PageController(initialPage: widget.index);
    _currentIndex.value = widget.index;
  }

  @override
  void dispose() {
    super.dispose();
    pageController.dispose();
  }

//##############################################################################

  void _handleIndexChanged(int index) {
    if (_currentIndex.value != index) {
      pageController
          .jumpToPage(_currentIndex.value < index ? index - 1 : index + 1);
      pageController.animateToPage(index,
          duration: const Duration(milliseconds: 700),
          curve: Curves.easeInOutQuart);
      _currentIndex.value = index;
    }
  }

//##############################################################################

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Container(
        color: const Color(0xFF131313),
        child: Row(
          children: [
            ValueListenableBuilder(
              valueListenable: _currentIndex,
              builder: (context, dynamic currentIndex, _) => ResponsiveBar(
                useIconScaleCurve: true,
                barCircularity: 0,
                iconPadding: const EdgeInsets.only(bottom: 25, top: 25),
                barThickness: 5,
                isForDesktop: true,
                currentIndex: currentIndex,
                onTap: (index) => _handleIndexChanged(index),
                iconSize: 35,
                scaling: 60,
                iconScaleAnimationFactor: 0.5,
                iconScaleCurve: Curves.decelerate,
                barSizeCurve: Curves.decelerate,
                barAccentColor: Colors.white,
                backgroundColor: Colors.black45,
                activeItemColor: Colors.white,
                passiveItemColor: Colors.white30,

                transition: const Duration(milliseconds: 400),
                itemsStaked: true,
                usingItemDescription: true,
                tooltipStayDuration: const Duration(milliseconds: 500),
                items: [
                  ResponsiveBarItem(
                      description: "test",
                      icon: const Icon(Icons.store_outlined),
                      selectedIcon: const Icon(Icons.store_rounded)),
                  ResponsiveBarItem(
                      description: "test2",
                      icon: const Icon(Icons.badge_outlined),
                      selectedIcon: const Icon(Icons.badge_rounded)),
                  ResponsiveBarItem(
                      description: "test3",
                      icon: const Icon(Icons.vpn_key_outlined),
                      selectedIcon: const Icon(Icons.vpn_key_rounded)),
                  ResponsiveBarItem(
                      description: "test4",
                      icon: const Icon(Icons.source_outlined),
                      selectedIcon: const Icon(Icons.source_rounded)),
                  ResponsiveBarItem(
                      description: "test5",
                      icon: const Icon(Icons.linear_scale_outlined),
                      selectedIcon: const Icon(Icons.linear_scale_rounded))
                ],
              ),
            ),
            Expanded(
              child: PageView(
                physics: const NeverScrollableScrollPhysics(),
                scrollDirection: Axis.vertical,
                controller: pageController,
                onPageChanged: (index) => _currentIndex.value = index,
                children: const <Widget>[
                  Placeholder(),
                  Placeholder(),
                  Placeholder(),
                  Placeholder(),
                  Placeholder(),
                ],
              ),
            )
          ],
        ),
      ),
    );
  }
}

class MobileMain extends StatefulWidget {
  MobileMain({super.key, required this.title, this.index = 0});

  final String title;
  int index = 0;

  @override
  State<MobileMain> createState() => _MobileMainState();
}

class _MobileMainState extends State<MobileMain> {
  final ValueNotifier<int> _currentIndex = ValueNotifier(0);
  late PageController pageController;

  @override
  void initState() {
    super.initState();
    pageController = PageController(initialPage: widget.index);
    _currentIndex.value = widget.index;
  }

  @override
  void dispose() {
    super.dispose();
    pageController.dispose();
  }

//##############################################################################

  void _handleIndexChanged(int index) {
    if (_currentIndex.value != index) {
      pageController
          .jumpToPage(_currentIndex.value < index ? index - 1 : index + 1);
      pageController.animateToPage(index,
          duration: const Duration(milliseconds: 700),
          curve: Curves.easeInOutQuart);
      _currentIndex.value = index;
    }
  }

//##############################################################################

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        body: Container(
      color: const Color(0xFF131313),
      child: Column(
        children: [
          Expanded(
            //################################################################
            //PageView
            //################################################################
            child: PageView(
              physics: const NeverScrollableScrollPhysics(),
              scrollDirection: Axis.vertical,
              controller: pageController,
              onPageChanged: (index) => _currentIndex.value = index,
              children: const <Widget>[
                Placeholder(),
                Placeholder(),
                Placeholder(),
                Placeholder(),
                Placeholder(),
              ],
            ),
          ),
          ValueListenableBuilder(
            valueListenable: _currentIndex,
            builder: (context, dynamic currentIndex, _) => ResponsiveBar(
              useBarAnimation: true,
              iconPadding: const EdgeInsets.all(13),
              usingTopBar: true,
              isForDesktop: false,
              currentIndex: currentIndex,
              onTap: _handleIndexChanged,
              iconSize: 30,
              scaling: 60,
              transition: const Duration(milliseconds: 500),
              iconScaleAnimationFactor: 0.5,
              iconScaleCurve: Curves.easeOutSine,
              barSizeCurve: Curves.decelerate,
              barAccentColor: Colors.white,
              backgroundColor: Colors.black38,
              activeItemColor: Colors.white,
              passiveItemColor: Colors.white38,
              items: [
                //######################################################
                //BarItems
                //######################################################
                ResponsiveBarItem(
                    icon: const Icon(Icons.store_outlined),
                    selectedIcon: const Icon(Icons.store_rounded)),
                ResponsiveBarItem(
                    icon: const Icon(Icons.badge_outlined),
                    selectedIcon: const Icon(Icons.badge_rounded)),
                ResponsiveBarItem(
                    icon: const Icon(Icons.vpn_key_outlined),
                    selectedIcon: const Icon(Icons.vpn_key_rounded)),
                ResponsiveBarItem(
                    icon: const Icon(Icons.source_outlined),
                    selectedIcon: const Icon(Icons.source_rounded)),
                ResponsiveBarItem(
                    icon: const Icon(Icons.linear_scale_outlined),
                    selectedIcon: const Icon(Icons.linear_scale_rounded))
              ],
            ),
          ),
        ],
      ),
    ));
  }
}
